linked_list/linked_list.py

```diff
@@ -1,128 +1,236 @@
+        self.head = None  # keep the head private. Not accessible outside this class
+        # self.end = None
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
-
+        if self.head == None:
+            return None
+        else:
+            return self.head.value
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        new_node = Node(value, next_node=self.head)
+        self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+
     def search(self, value):
-        pass
+        current = self.head
+
+        while current:
+            if current.value == value:
+                return True
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+
     def length(self):
-        pass
+        current = self.head
+        length = 0
+
+        while current:
+            length += 1
+            current = current.next
+        return length
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        current = self.head
+        current_index = 0
+
+        while current:
+            if current_index == index:
+                return current.value
+            current_index += 1
+            current = current.next
+        if current_index < index:
+            return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+
     def get_last(self):
-        pass
+        current = self.head
+
+        while current:
+            if current.next == None:
+                return current.value
+            current = current.next
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_last(self, value):

```

Comment:
👍   However the time complexity is O(n) because the way you wrote this you have to traverse the entire list until the very end.

You **could** write in a `self.tail` reference which always refers to the last node in the list.  Then with some additional code, you could maintain O(1) `add_last` and `get_last` functions.